HLTrigger/Muon/plugins/HLTMuonL3orL2PreFilter.cc

```diff
@@ -0,0 +1,390 @@
+      float vz_L2 = L2cand->vz() - beamSpot.z0();
+      float vx_L2 = L2cand->vx() - beamSpot.x0();
+      float vy_L2 = L2cand->vy() - beamSpot.y0();
+      if (std::abs(vz_L2 - (vx_L2 * L2cand->px() + vy_L2 * L2cand->py()) / (L2cand->pt() * L2cand->pt()) *
+                               L2cand->pz()) > max_Dz_)
+        FailedL2 = true;
+
+      // dxy significance cut (safeguard against bizarre values)
+      float const absDxy = std::abs(L2tk->dxy(beamSpot.position()));
+      if (min_DxySig_ > 0 && (L2tk->dxyError() <= 0 || (absDxy < min_DxySig_ * L2tk->dxyError())))
+        FailedL2 = true;
+
+      //dxy beamspot cut
+      if (absDxy > max_DXYBeamSpot_ || absDxy < min_DXYBeamSpot_)
+        FailedL2 = true;
+    }
+
+    //min muon hits cut
+    const reco::HitPattern& L2trackHits = L2tk->hitPattern();
+    if (L2trackHits.numberOfValidMuonHits() < min_NmuonHits_L2_)
+      FailedL2 = true;
+
+    // number of eta bins for cut on number of stations
+    const std::vector<double>::size_type nAbsetaBins = absetaBins_L2_.size();
+
+    // cut on number of stations
+    bool failNstations(false), failNhits(false), failNchambers(false);
+    for (unsigned int i = 0; i < nAbsetaBins; ++i) {
+      if (std::abs(L2cand->eta()) < absetaBins_L2_[i]) {
+        if (L2cand->track()->hitPattern().muonStationsWithAnyHits() < minNstations_L2_[i]) {
+          failNstations = true;
+        }
+        if (L2cand->track()->numberOfValidHits() < min_Nhits_L2_[i]) {
+          failNhits = true;
+        }
+        if ((L2cand->track()->hitPattern().dtStationsWithAnyHits() +
+                 L2cand->track()->hitPattern().cscStationsWithAnyHits() <
+             minNchambers_L2_[i])) {
+          failNchambers = true;
+        }
+        break;
+      }
+    }
+    if (failNstations || failNhits || failNchambers)
+      FailedL2 = true;
+
+    //pt difference cut
+    double L2candPt = L2cand->pt();
+    double L2trackPt = L2tk->pt();
+
+    if (std::abs(L2candPt - L2trackPt) > max_PtDifference_)
+      FailedL2 = true;
+
+    //track pt cut
+    if (L2trackPt < min_TrackPt_)
+      FailedL2 = true;
+
+    // Pt threshold cut
+    double L2pt = L2cand->pt();
+    double L2err0 = L2tk->error(0);
+    double L2abspar0 = std::abs(L2tk->parameter(0));
+    double L2ptLx = L2pt;
+    // convert 50% efficiency threshold to 90% efficiency threshold
+    if (L2abspar0 > 0)
+      L2ptLx += nsigma_Pt_ * L2err0 / L2abspar0 * L2pt;
+    LogTrace("HLTMuonL3orL2PreFilter") << " ...Muon in loop, trackkRef pt= " << L2tk->pt() << ", ptLx= " << L2ptLx
+                                       << " cand pT " << L2cand->pt();
+    if (L2ptLx < min_Pt_L2_)
+      FailedL2 = true;
+
+    bool matchedL3 = false;
+
+    ////////////////////////////////////////L3///////////////////////////////////////////////////////////
+    for (unsigned int iMu = 0; iMu < mucands.size(); iMu++) {
+      reco::RecoChargedCandidateRef cand(iEvent.getHandle(candToken_), iMu);
```

Comment:
```suggestion
      reco::RecoChargedCandidateRef cand(iEvent.getHandle(candToken_), iMu);
```
The `getHandle` here is basically a repetition of what was used to initialise `mucands`. Since the handle is needed, you can use `auto const mucands = iEvent.getHandle(candToken_)` upstairs, and then replace `mucands.` with `mucands->` throughout, plus replace here `iEvent.getHandle(candToken_)` with `mucands`.